runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,365 @@
+      for row = start_row, end_row do
+        local level = row_level[row] or { 0 }
+        level[1] = level[1] + 1
+        row_level[row] = level
+      end
+      row_level[start_row][2] = '>'
+      row_level[end_row][2] = '<'
+    end
+  end
+
+  bufstate.row_level = row_level
+  bufstate.row_text = row_text
+  bufstate.kind_ranges = kind_ranges
+  bufstate.version = ctx.version
+
+  if api.nvim_get_mode().mode:match('^i') then
+    -- `foldUpdate()` is guarded in insert mode.
+    schedule_foldupdate(bufnr)
+  else
+    foldupdate(bufnr)
+  end
+end
+
+---@param bufnr integer
+---@param client vim.lsp.Client The client whose server supports `foldingRange`.
+---@return integer? request_id
+local function refresh(bufnr, client)
+  ---@type lsp.FoldingRangeParams
+  local params = { textDocument = util.make_text_document_params(bufnr) }
+  local _, request_id = client:request(ms.textDocument_foldingRange, params, handler, bufnr)
+  return request_id
+end
+
+-- MAINTENANCE NOTE:
+-- * `bufstate` requires event hooks to ensure correctness.
+-- * event hooks requires the previous `bufstate` for updates.
+--
+-- In other words, they are interdependent.
+-- Since their creation and destruction are handled manually,
+-- our strategy is to ensure that their lifecycles are always synchronized.
+--
+-- As they are only meaningful for buffers associated with
+-- at least one window that have `foldexpr()` set,
+-- the following three elements are guaranteed
+-- to exist or not exist simultaneously:
+--
+-- * A buffer associated with at least one window where `foldexpr()` is set.
+-- * Its `bufstate`.
+-- * Its event hooks.
+local augroup_setup = api.nvim_create_augroup('vim_lsp_folding_range/setup', {})
+
+--- Setup `bufstate` and its event hooks.
+---@param bufnr integer
+---@return vim.lsp.folding_range.BufState
+local function setup(bufnr)
+  ---@type vim.lsp.folding_range.BufState
+  local bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+  if not api.nvim_buf_is_loaded(bufnr) then
+    return bufstate
+  end
+
+  -- Register `bufstate`.
+  bufstates[bufnr] = bufstate
+
+  -- Setup event hooks.
+  api.nvim_buf_attach(bufnr, false, {
+    on_detach = function()
+      bufstates[bufnr] = nil
+    end,
+    on_reload = function()
+      bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+      bufstates[bufnr] = bufstate
+      for _, client in
+        ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+      do
+        refresh(bufnr, client)
+      end
+    end,
+    --- When modifying the buffer, new foldinfo is scheduled to be updated,
+    --- synchronize the changed rows with their previous foldlevels as possible,
+    --- before new foldlevels is updated.
+    ---@param start_row integer
+    ---@param old_row integer
+    ---@param new_row integer
+    on_bytes = function(_, _, _, start_row, _, _, old_row, _, _, new_row, _, _)
+      if bufstate == nil then
+        return true
```

Comment:
I think you can remove the autocmds heres.

Reply:
As I noted in the code above, the lifecycle of bufstate and these callbacks should be synchronized. So if we do not need to destroy bufstate here, then we also do not need to destroy the callback. However, callbacks of `nvim_buf_attach` can only be destroyed if it returns true, which is the reason for adding the branch here.